src/DataTypes/Utils.cpp

```diff
@@ -0,0 +1,228 @@
+
+                return false;
+            }
+            case TypeIndex::Int8:
+            case TypeIndex::Int16:
+            case TypeIndex::Int32:
+            case TypeIndex::Int64:
+            case TypeIndex::Int128:
+            case TypeIndex::Int256: {
+                if (rhs_which_type.isInt())
+                    return rhs_type->getSizeOfValueInMemory() >= lhs_type->getSizeOfValueInMemory();
+
+                if (rhs_which_type.isString())
+                    continue;
+
+                return false;
+            }
+            case TypeIndex::Float32: {
+                if (rhs_which_type.isFloat64() || rhs_which_type.isString())
+                    continue;
+
+                return false;
+            }
+            case TypeIndex::Float64:
+            case TypeIndex::Date:
+            case TypeIndex::Date32:
+            case TypeIndex::DateTime:
+            case TypeIndex::DateTime64:
+            case TypeIndex::FixedString:
+            case TypeIndex::Enum8:
+            case TypeIndex::Enum16:
+            case TypeIndex::IPv6:
+            {
+                if (rhs_which_type.isString())
+                    continue;
+
+                return false;
+            }
+            case TypeIndex::Decimal32:
+            case TypeIndex::Decimal64:
+            case TypeIndex::Decimal128:
+            case TypeIndex::Decimal256:
+            {
+                if (rhs_which_type.isDecimal())
+                {
+                    auto lhs_type_decimal_precision = getDecimalPrecision(*lhs_type);
+                    auto rhs_type_decimal_precision = getDecimalPrecision(*rhs_type);
+                    if (lhs_type_decimal_precision > rhs_type_decimal_precision)
+                        return false;
+
+                    auto lhs_type_decimal_scale = getDecimalScale(*lhs_type);
+                    auto rhs_type_decimal_scale = getDecimalScale(*rhs_type);
+                    if (lhs_type_decimal_scale > rhs_type_decimal_scale)
+                        return false;
+
+                    continue;
+                }
+
+                if (rhs_which_type.isString())
+                    continue;
+
+                return false;
+            }
+            case TypeIndex::UUID:
+            {
+                if (rhs_which_type.isString() || rhs_which_type.isUInt128())
+                    continue;
+
+                return false;
+            }
+            case TypeIndex::IPv4:
+            {
+                if (rhs_which_type.isUInt32() || rhs_which_type.isUInt64() || rhs_which_type.isString())
+                    continue;
+
+                return false;
+            }
+            case TypeIndex::Nullable:
+            {
+                if (rhs_type_was_nullable)
+                {
+                    const auto & lhs_type_nullable = assert_cast<const DataTypeNullable &>(*lhs_type);
+                    return canBeSafelyCasted(lhs_type_nullable.getNestedType(), rhs_type);
```

Comment:
This is not correct because the stack may not be empty.
Looks like `Map(Nullable(x), y)` is always equal `Map(Nullable(x), z)` now.

Btw I think we can always use recursion here. A ton of other code uses it to traverse types.

Reply:
Updated.